smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/traits/StaticContextParamsTrait.java

```diff
@@ -0,0 +1,140 @@
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeMapper;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.AbstractTrait;
+import software.amazon.smithy.model.traits.AbstractTraitBuilder;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.BuilderRef;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+/**
+ * Binds static values to AWS endpoint parameters for the targeted operation.
+ */
+@SmithyUnstableApi
+public final class StaticContextParamsTrait extends AbstractTrait implements ToSmithyBuilder<StaticContextParamsTrait> {
+    public static final ShapeId ID = ShapeId.from("smithy.rules#staticContextParams");
+
+    private final Map<String, StaticContextParamDefinition> parameters;
+
+    private StaticContextParamsTrait(Builder builder) {
+        super(ID, builder.getSourceLocation());
+        this.parameters = builder.parameters.copy();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public Map<String, StaticContextParamDefinition> getParameters() {
+        return parameters;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+        StaticContextParamsTrait that = (StaticContextParamsTrait) o;
+        return getParameters().equals(that.getParameters());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), getParameters());
+    }
+
+    @Override
+    protected Node createNode() {
+        NodeMapper mapper = new NodeMapper();
+        mapper.setOmitEmptyValues(true);
+        return mapper.serialize(getParameters()).expectObjectNode();
+    }
+
+    @Override
+    public SmithyBuilder<StaticContextParamsTrait> toBuilder() {
+        return new Builder()
+                .sourceLocation(getSourceLocation())
+                .parameters(parameters);
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            NodeMapper mapper = new NodeMapper();
+            Map<String, StaticContextParamDefinition> parameters = new LinkedHashMap<>();
+            value.expectObjectNode().getMembers().forEach((stringNode, node) -> {
+                parameters.put(stringNode.getValue(), mapper.deserialize(node, StaticContextParamDefinition.class));
+            });
+            return builder()
```

Comment:
Would be good to set the nodeCache here too. Example: https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/traits/HttpTrait.java#L60